File: week1/java/week1_3.java
package week1;

import java.util.ArrayList;

public class week1_3 {
    public static int answer = 0;
    public static boolean[] v;
    public static ArrayList<Integer> a = new ArrayList<>();
    static boolean isPrime(int num){
        if(num == 1 || num == 0) return false;
        for(int i = 2; i*i <= num; i++){
            if(num % i == 0) return false;
        }
        return true;
    }

    public static void go(char[] number, String str) {
        if(str.length() > number.length) return ;

        int n = 0;
        for(int i=0; i<str.length(); i++) {
            n += (str.charAt(i)-'0') * ((int) Math.pow(10, str.length()-1-i));
        }
        if(isPrime(n)) {
            boolean re = true;
            for(int i=0; i<a.size(); i++){
                if(a.get(i) == n){
                    re = false;
                    break;
                }
            }
            if(re){
                a.add(n);
                answer += 1;
            }
        }
        String temp = str;
        for(int i = 0; i < number.length; i++){
            if(v[i]==false){
                v[i] = true;
                str = temp + number[i];
                if(str.charAt(0) != '0') go(number, str);
                v[i] =false;
            }
        }
    }
    public static int solution(String numbers){
        char[] number = new char[numbers.length()];
        v = new boolean[numbers.length()];
        for(int i=0; i<numbers.length(); i++) {
            number[i] = numbers.charAt(i);
        }
        go(number, "");
        return answer;
    }
    public static void main(String[] args){
        System.out.println(solution("011"));
    }
}

File: week1/c++/week1_3.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool isPrime(int num)
{
    if (num < 2)
    {
        return false;
    }

    for (int i = 2; i * i <= num; i++)
    {
        if (num % i == 0)
        {
            return false;
        }
    }

    return true;
}

bool isNumber(int num, vector<int> arr)
{

    string n = to_string(num);
    vector<int> m;
    for (int i = 0; i < n.size(); i++)
    {
        m.push_back(n[i] - '0');
    }

    bool check;
    for (int i = 0; i < m.size(); i++)
    {
        check = false;

        for (int j = 0; j < arr.size(); j++)
        {
            if (m[i] == arr[j])
            {
                check = true;
                arr[j] = -1;
                break;
            }
        }

        if (check == false)
        {
            return false;
        }
    }

    return check;
}

int solution(string numbers)
{

    string max;
    max = numbers;
    sort(max.begin(), max.end(), greater<int>());

    int answer = 0;

    vector<int> n;
    for (int i = 0; i < numbers.size(); i++)
    {
        n.push_back(numbers[i] - '0');
    }

    for (int i = 2; i <= stoi(max); i++)
    {
        if (isPrime(i))
        {
            if (isNumber(i, n))
            {
                answer++;
            }
        }
    }

    return answer;
}